builder/comp-builder.nix

```
@@ -23,6 +23,14 @@ let self =
 , preInstall ? component.preInstall , postInstall ? component.postInstall
 , preHaddock ? component.preHaddock , postHaddock ? component.postHaddock
 , shellHook ? ""
+, configureAllComponents ? component.configureAllComponents ||
+    # When set, configure all the components in the package
+    # (not just the one we are building).
+    # Enable for tests in packages that use cabal-doctest.
+    ( haskellLib.isTest componentId &&
+      lib.any (x: x.identifier.name or "" == "cabal-doctest") package.setup-depends
```

I think `package.setup-depends` is wrong here. If a single `test-suite` depends on `cabal-doctest` then `setup-depends` will contain `cabal-doctest`. But if _another_ `test-suite` is defined that *doesn't* depend on `cabal-doctest`, then this `test-suite` will _also_ build all components. I think this is wrong. Can we look at just what this component depends on?

---

Here's my humble suggestion:

```suggestion
      lib.any (x: x.identifier.name or "" == "cabal-doctest") package.setup-depends &&
      lib.any (x: x.identifier.name or """ == "doctest") component.depends
```

---

Dumb suggestion: what if we just made people set this themselves? This is only a problem for test components using `doctest`, so is it totally unreasonable to say "if you use doctest, you should set this option on the corresponding test component"?

---

I guess we could, but I like the mantra of "if you do `cabal X` and it works, then it will work with Haskell.nix". That will no longer hold for this case now.

---

I have a feeling I'm about to say something I've said in the past, but... why does this work in cabal itself? I vaguely recall that cabal turns off per-component builds if you have a custom setup - should we do the same thing? Maybe that would work more simply than trying to detect `cabal-doctest`?

---

That one is outside my knowledge, I'm afraid!

---

> should we do the same thing?

I think this change is a step in that direction.

---

I don't think it works well enough to use for all custom setup packages yet.  In particular I think the component filtering will not work.